### ISSUE TITLE
exfat analyzer failed sometimes

### ISSUE THREAD

**Issue by SuperGod** [sleuthkit/sleuthkit#651]
I use fls.exe to test.This time it success:

![image](https://cloud.githubusercontent.com/assets/4470750/15347958/daf9b99c-1cf8-11e6-8963-850f8fe973be.png)

output is:

```
r/r 6147:       Super (Volume Label Entry)
r/r 6148:       $ALLOC_BITMAP
r/r 6149:       $UPCASE_TABLE
v/v 2000562915: $MBR
v/v 2000562916: $FAT1
d/d 2000562917: $OrphanFiles
```

But I change the arg to 128k,and then it failed.

![image](https://cloud.githubusercontent.com/assets/4470750/15347973/fa5a1c32-1cf8-11e6-9e6d-bcccf1f277f9.png)

output error is:

```
Invalid magic value (Error: sector size (0) is not a multiple of device size (51
2)
Do you have a disk image instead of a partition image?)
```

**Comment by SuperGod** [comment]
I just change the "csize" define in "struct  FATFS_INFO" in tsk_fatfs.h, and it seems it work fine,but I do not know this change is real correct:

```
struct FATFS_INFO {
        TSK_FS_INFO fs_info;    /* super class */

        /* FAT cache */
        /* cache_lock protects fatc_buf, fatc_addr, fatc_ttl */
        tsk_lock_t cache_lock;
        char fatc_buf[FATFS_FAT_CACHE_N][FATFS_FAT_CACHE_B];        //r/w shared - lock
        TSK_DADDR_T fatc_addr[FATFS_FAT_CACHE_N];     // r/w shared - lock
        uint8_t fatc_ttl[FATFS_FAT_CACHE_N];  //r/w shared - lock

        /* First sector of FAT */
        TSK_DADDR_T firstfatsect;

        /* First sector after FAT  - For TSK_FS_INFO_TYPE_FAT_12 and TSK_FS_INFO_TYPE_FAT_16, this is where the
         * root directory entries are.  For TSK_FS_INFO_TYPE_FAT_32, this is the the first
         * cluster */
        TSK_DADDR_T firstdatasect;

        /* The sector number were cluster 2 (the first one) is
         * for TSK_FS_INFO_TYPE_FAT_32, it will be the same as firstdatasect, but for TSK_FS_INFO_TYPE_FAT_12 & 16
         * it will be the first sector after the Root directory  */
        TSK_DADDR_T firstclustsect;

        /* size of data area in clusters, starting at firstdatasect */
        TSK_DADDR_T clustcnt;

        TSK_DADDR_T lastclust;

        /* sector where the root directory is located */
        TSK_DADDR_T rootsect;

        uint32_t dentry_cnt_se; /* max number of dentries per sector */
        uint32_t dentry_cnt_cl; /* max number of dentries per cluster */

        uint16_t ssize;         /* size of sectors in bytes */
        uint16_t ssize_sh;      /* power of 2 for size of sectors.  >> to divide by sector size.  << to multiply by sector size */
        uint32_t csize;          /* size of clusters in sectors */
        uint8_t numfat;         /* number of fat tables */
        uint32_t sectperfat;    /* sectors per fat table */
        uint16_t numroot;       /* number of 32-byte dentries in root dir */
        uint32_t mask;          /* the mask to use for the sectors */

        TSK_INUM_T mbr_virt_inum;
        TSK_INUM_T fat1_virt_inum;
        TSK_INUM_T fat2_virt_inum;

        tsk_lock_t dir_lock;    //< Lock that protects inum2par.
        void *inum2par;         //< Maps subfolder metadata address to parent folder metadata addresses.

		char boot_sector_buffer[FATFS_MASTER_BOOT_RECORD_SIZE];
        int using_backup_boot_sector;

		TSK_FATFS_SUBTYPE_ENUM subtype; // Identifies any variations on the standard FAT format

        int8_t (*is_cluster_alloc)(FATFS_INFO *fatfs, TSK_DADDR_T clust);

        uint8_t (*is_dentry)(FATFS_INFO *a_fatfs, FATFS_DENTRY *a_dentry,
            FATFS_DATA_UNIT_ALLOC_STATUS_ENUM a_cluster_is_alloc,
            uint8_t a_do_basic_tests_only);

        uint8_t (*inode_lookup)(FATFS_INFO *a_fatfs, TSK_FS_FILE *a_fs_file,
            TSK_INUM_T a_inum);

        uint8_t (*inode_walk_should_skip_dentry)(FATFS_INFO *a_fatfs,
            TSK_INUM_T a_inum, FATFS_DENTRY *a_dentry,
            unsigned int a_selection_flags, int a_cluster_is_alloc);

        uint8_t (*istat_attr_flags) (FATFS_INFO *a_fatfs, TSK_INUM_T a_inum,  FILE *a_hFile);

        TSK_RETVAL_ENUM (*dent_parse_buf)(FATFS_INFO *a_fatfs,
            TSK_FS_DIR *a_fs_dir, char *a_buf, TSK_OFF_T a_buf_len,
            TSK_DADDR_T *a_sector_addrs);

        TSK_RETVAL_ENUM (*dinode_copy)(FATFS_INFO *a_fatfs, TSK_INUM_T a_inum,
            FATFS_DENTRY *a_dentry, uint8_t a_cluster_is_alloc, TSK_FS_FILE *a_fs_file);

        struct {
            uint64_t first_sector_of_alloc_bitmap;
            uint64_t length_of_alloc_bitmap_in_bytes;
        } EXFATFS_INFO;
	};
```

I test of this:
![image](https://cloud.githubusercontent.com/assets/4470750/15348655/24b2fe12-1cff-11e6-9b3b-463e0591cc1e.png)

and the output is:

```
r/r 2097155:    Super (Volume Label Entry)
r/r 2097156:    $ALLOC_BITMAP
r/r 2097157:    $UPCASE_TABLE
v/v 1998629603: $MBR
v/v 1998629604: $FAT1
d/d 1998629605: $OrphanFiles
```

it seems all ok.

**Comment by bcarrier** [comment]
Hello.  What did you change between the first two runs (I cannot read it).  Was it the cluster size that you changed from 64KB to 128KB?

When you say you changed csize, was it that you just changed it from a uint8_t to a uint32_t?

I can certainly see that csize should be bigger than 8-bits to handle large ExFAT images (I think only 16-bits is needed though to support the 32MB clusters).

I think what was happening was that csize was getting set to 0 because it was only the lower 8-bits and that screwed up follow on calculations. 

This fix is in the release-4.3.0 branch.

**Comment by lfcnassif** [comment]
Hi Brian, I think more than 16 bits are needed to handle 32MB clusters, as more than 8bits are needed to handle 128KB clusters.

**Comment by bcarrier** [comment]
Actually, you're right.  We need 17 bits, which means we should bump it to 32.

Max cluster size is 32MB and min sector size is 512.  So the biggest number of sectors per cluster will be 65,536. 

Fixed with 8b5182bc84d40ed215e7000411fc027dac244af8.
